### DISCUSSION TITLE
(Port-)forwarding/Page preview

### DISCUSSION THREAD

**Discussion by janus-reith** [eclipse-theia/theia#7822] [General]
Many of my projects involve multiple services sitting on multiple ports.
For example, I have a frontend on port 4000,  an api on port 3000, and some oauth flow using port 5000.

How could I forward these in the browser using theia? Preferrably I would not want to  use some separate ssh forwarding that has to run on my local machine.

But at the same time, I would want my scripts and dom being executed and evaluated locally in the browser, so I could inspect things properly.
Another consideration would be, that things like the oauth flow should be functional, so I assume some kind of rewrite for localhost requests would need to happen.

I'm currently using VS Code with Remote SSH plugin and some ssh LocalForwards set up for several servers each containing separate projects, and I'm trying to find ways to replace this setup.

Any idea how this could work out?
Thanks!

**Comment by akosyakov** [comment]
@janus-reith please see: https://github.com/eclipse-theia/theia/blob/96e3c728476ec0fb0982ae19580ea347c79f97a2/packages/core/src/browser/external-uri-service.ts#L25-L33 you can customize it, default behaviour is documented in js-doc

**Comment by akosyakov** [comment]
Unfortunately, it works for everything besides the mini-browser, we have an issue for it: https://github.com/eclipse-theia/theia/issues/6562

For the mini-browser maybe @kittaakos can give a hint how forwarding of localhost can be implemented

**Reply by kittaakos** [reply]
You can try to customize the `LocationWithoutSchemeMapper` or the implementation of other location mappers. You might want to check [this](https://github.com/eclipse-theia/theia/pull/4357#issuecomment-465592237) conversation.

**Comment by janus-reith** [comment]
Thanks @akosyakov @kittaakos for the details!

I will need to dive further into this, as currently I'm struggling to understand it properly: If the mini-browser, which I assumme is the kind of preview browser I was looking for, currently does not work with url mapping, then where does that url mapper actually come into play? Or, is there some other pewview browser besides that mini-browser?

**Reply by akosyakov** [reply]
Links to localhost in terminals, editors, VS Code webviews and so on.
